### ISSUE TITLE
Fix internal methods that only target platform objects w/ indexed/named props

### ISSUE THREAD

**Issue by tobie** [heycam/webidl#202]
The algorithm's introductory prose incorrectly limits the algorithm streps to platform objects that support indexed/named props, for example:

> The internal [[Delete]] method of every platform object O that implements an interface which supports indexed or named properties must behave as follows when called […].

Instead, the method should apply to _all_ platform objects and see its algorithm steps forked to address the cases of platform objects which support indexed or named properties.

In addition, the common case should rely on the relevant ES `Ordinary[X]` method rather than a custom default behavior solution.

The following methods are affected:
- [[Set]]
- [[DefineOwnProperty]]
- [[Delete]]
- [[Call]]

**Comment by bzbarsky** [comment]
Why is this better than the current setup?

The current setup makes it obvious that for objects without indexed or named props no magic needs to happen and no magic _can_ happen.  That makes it less likely that a spec edit will accidentally introduce magic for them.

> In addition, the common case should rely on the relevant ES Ordinary[X] method rather than a custom default behavior solution.

I'm not sure what you mean here.  If you mean that steps 3 and 4 of https://heycam.github.io/webidl/#delete should be replaced with a call to https://tc39.github.io/ecma262/#sec-ordinarydelete then it's not obvious to me that this is black-box identical, but maybe it is.  It would need to be carefully reasoned through.  Same for the other internal methods.

**Comment by domenic** [comment]
I think I gave some bad advice here. I thought that the current setup was contradictory, since the intro paragraphs say "this only applies for objects that support named/indexed props", but then there are fallback steps outside of those for named/indexed props. But of course now I realize those steps are for the rest of the properties of the object (the non-named/indexed ones). So the current spec is not broken, like I thought it was and told @tobie it was.

I would still prefer a clearer setup, however, where there are _n_ types of exotic objects, and they all have internal method definitions that apply in every case, not just sometimes. A simple way to do this is to have n = 1, with the only type of exotic object being "platform object", and then we just define all the internal methods, with some if statements inside them. I guess you could go for n = 2 (platform objects and named-or-indexed platform objects), or even n = 4 (those two multiplied by legacy-caller-or-not).

**Comment by annevk** [comment]
I think we should aim for any new platform objects to not require a proxy.

**Comment by tobie** [comment]
Just to clarify: my initial understanding that these methods (in 3.8) are only defined for platform objects that support indexed/named props is correct, right? Other platform objects that don't support index/named props just use whatever [[X]] method is defined in ECMAScript, right?

So for consistency, shouldn't the platform object [[[PreventExtensions]] method](https://heycam.github.io/webidl/#preventextensions), which we just added, be redefined as such:

```
The internal [[PreventExtensions]] method of every platform object that
implements an interface which supports indexed or named properties,
must behave as follows:

1. return false.
```

**Comment by domenic** [comment]
@tobie yeah, I was confused and gave you bad advice about PreventExtensions. I'm saying I would kind of prefer the others to look like PreventExtensions does, but we don't seem to have agreement on that.

**Comment by annevk** [comment]
I think the n = 2 approach is fine. Indexed/named getters are just as legacy as callers, imo. "Platform objects" (no proxy needed) vs "legacy platform objects" (proxy needed).

**Comment by bzbarsky** [comment]
> https://github.com/heycam/webidl/issues/202#issuecomment-256049412

That's fair. 

> I guess you could go for n = 2 (platform objects and named-or-indexed platform objects), or even n = 4 (those two multiplied by legacy-caller-or-not)

legacycaller is used in three places in Gecko: HTMLObjectElement, HTMLEmbedElement, HTMLAllCollection.  Those are also the only places it's used in the HTML spec.  I'm about to remove the former two (see <https://bugzilla.mozilla.org/show_bug.cgi?id=909656>), because we have near-zero telemetry (2 hits out of 460+ million telemetry pings) _and_ we just dropped support for non-Flash plugins and I'm fairly certain Flash doesn't use this stuff.  I filed https://github.com/whatwg/html/issues/1971 to remove from the spec too.  Of course HTMLAllCollection already has indexed/named props.

So I think we should do n == 2, change IDL to require that legacycaller is only specified if indexed or named props are present, and then we only need interesting internal methods for the indexed/named prop case.  We can continue hating HTMLAllCollection, of course.  ;)

**Comment by tobie** [comment]
So I'm nearly finished with a PR for this.

I started off (https://github.com/tobie/webidl/commit/9e54e64fe575a5959c7d0601b11609c7cd35ff64) considering legacy platform objects and platform objects as two disjoint sets. This had the benefit of forcing me to make explicit decisions as to whether each occurence of the term "platform object" referenced all platform objects or just legacy or non-legacy ones.

It turns out that there are very few places where we want to refer exclusively to platform objects which are not legacy ones, so the end result of this approach is to litter the spec with the following construct: 

```
[=platform object|platform=] or [=legacy platform object|legacy platform=] objects
```

Having gone down that road, I think the right approach is to consider legacy platform objects as a proper subset of platform objects and, in the rare cases where that's really needed, refer to the complement of legacy platform objects as: "platform objects which are not legacy platform objects." 

I'll amend the branch accordingly unless I hear strong voices in favor of the disjoint sets approach.

**Comment by annevk** [comment]
How can legacy be a subset? Only legacy has getters/setters/callers.

**Comment by tobie** [comment]
Am I getting set theory all mixed up (again)?

All legacy platform objects are also platform objects, but not all platform objects are legacy platform objects => legacy platform objects are a proper [subset](https://en.wikipedia.org/wiki/Subset) of platform objects, no?

**Comment by annevk** [comment]
I think they are a superset as they offer more features. Platform objects seems like the subset.

**Comment by tobie** [comment]
They do offer more features, yes. But if you take classical inheritance, you see the same pattern going on, no?

If `Cat` is a subclass of `Animal`, it's to add more feature to the the `Cat` class, such as the `.meow()` method. Yet cats are a (proper) subset of animals.

Similarly, when we currently say in the spec "all platform objects," this includes legacy platform objects, right? If we agree that this is what we want (and I think we do), we don't really need to agree on how that translates in set theory terminology for me to be able to move ahead with it.

**Comment by domenic** [comment]
@annevk, you are talking about the set of features, whereas @tobie is talking about the set of objects.

The set of legacy platform objects is a subset of the set of platform objects, if we define it @tobie's way. Which I think we should.

The set of features offered by legacy platform objects is a superset of the set of features offered by platform objects, if we define the set of platform objects as disjoint from the set of legacy platform objects.

As far as I can tell @tobie's setup will work well. We might need a third category, global platform objects. But in general non-legacy/non-global platform objects won't override any internal methods, which sounds great.

**Comment by annevk** [comment]
Sounds good, sorry for the confusion. Globals do indeed seem like a subclass of sorts too. Would be nice to have a shorthand for non-legacy-non-global, but maybe only once we identify a clear need.

**Comment by tobie** [comment]
POPO (Plain Old Platform Objects)? :D
